linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,455 @@
+                        {
+                            if (ThreadAPI_Create(&threadpool->thread_handle_list[threadpool->list_index], threadpool_work_func, threadpool) != THREADAPI_OK)
+                            {
+                                // Decrement the thread count since we didn't actually create the thread
+                                interlocked_decrement(&threadpool->thread_count);
+                                LogError("Failure creating thread");
+                                break;
+                            }
+                            else
+                            {
+                                sem_post(&threadpool->semaphore);
+                                threadpool->list_index++;
+                                result = 0;
+                                goto all_ok;
+                            }
+                        }
+                    }
+                    else
+                    {
+                        sem_post(&threadpool->semaphore);
+                        result = 0;
+                        goto all_ok;
+                    }
+                }
+                while(1);
+                result = MU_FAILURE;
+            }
+        }
+    }
+all_ok:
+    return result;
+}
+
+int threadpool_open_async(THREADPOOL_HANDLE threadpool, ON_THREADPOOL_OPEN_COMPLETE on_open_complete, void* on_open_complete_context)
+{
+    int result;
+    if (
+        (threadpool == NULL) ||
+        (on_open_complete == NULL))
+    {
+        LogError("THREADPOOL_HANDLE threadpool=%p, ON_THREADPOOL_OPEN_COMPLETE on_open_complete=%p, void* on_open_complete_context=%p",
+            threadpool, on_open_complete, on_open_complete_context);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        int32_t current_state = interlocked_compare_exchange(&threadpool->state, (int32_t)THREADPOOL_STATE_OPEN, (int32_t)THREADPOOL_STATE_NOT_OPEN);
+        if (current_state != (int32_t)THREADPOOL_STATE_NOT_OPEN)
+        {
+            LogError("Not closed, cannot open, current state is %" PRI_MU_ENUM "", MU_ENUM_VALUE(THREADPOOL_STATE, current_state));
+            result = MU_FAILURE;
+        }
+        else
+        {
+            on_open_complete(on_open_complete_context, THREADPOOL_OPEN_OK);
+            result = 0;
+        }
+    }
+    return result;
+}
+
+void threadpool_close(THREADPOOL_HANDLE threadpool)
+{
+    int result;
+    if (threadpool == NULL)
+    {
+        LogError("THREADPOOL_HANDLE threadpool=%p", threadpool);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        int32_t current_state = interlocked_compare_exchange(&threadpool->state, (int32_t)THREADPOOL_STATE_NOT_OPEN, (int32_t)THREADPOOL_STATE_OPEN);
+        if (current_state != (int32_t)THREADPOOL_STATE_NOT_OPEN)
+        {
+            LogError("Not closed, cannot open, current state is %" PRI_MU_ENUM "", MU_ENUM_VALUE(THREADPOOL_STATE, current_state));
+            result = MU_FAILURE;
+        }
```

Comment:
Shouldn't we wait for existing tasks to complete here?

Reply:
We are waiting in the destroy since open doesn't do anything either.